include/zephyr/printk-hooks.h

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_PRINTK_HOOKS_H_
+#define ZEPHYR_INCLUDE_PRINTK_HOOKS_H_
+
+/**
+ * @brief Install the character output routine for printk
+ *
+ * To be called by the platform's console driver at init time. Installs a
+ * routine that outputs one ASCII character at a time.
+ * @param fn putc routine to install
+ */
+void __printk_hook_install(int (*fn)(int c));
```

Comment:
One nitpick:  if you're promoting this to an API a `typedef int (*printk_hook_fn_t)(int c);` might look cleaner than a bare function pointer.

Reply:
Not a bad idea, added a patch for that.